pycsob/client.py

```diff
@@ -17,6 +21,208 @@ def send(self, request, **kwargs):
         return super(HTTPAdapter, self).send(request, **kwargs)
 
 
+def reorder_fields(instance: object) -> Iterable[Field]:
+    """Reorder fields of dataclass."""
+    for name in instance._ordered_fields:
+        yield instance.__dataclass_fields__[name]
+
+
+class ConvertMixin:
+    """Convert instance into ordered dict."""
+
+    _max_length: Dict[str, int] = {}
+
+    def _to_camel_case(self, name: str) -> str:
+        """Convert name to camel case form."""
+        parts = name.split("_")
+        return parts[0] + "".join(i.title() for i in parts[1:])
+
+    def _format_field(self, name: str, value: Any) -> str:
+        if isinstance(value, Enum):
+            value = str(value.value)
+        elif isinstance(value, (date, datetime)):
+            value = value.isoformat()
+        elif isinstance(value, str) and name in self._max_length:
+            value = value[:self._max_length[name]].rstrip()
+        else:
+            value = getattr(self, f"_format_{name}", lambda v: v)(value)  # Custom formatting method
+        return value
+
+    def to_dict(self) -> Dict[str, Union[str, Dict]]:
+        data = []
+        get_fields = reorder_fields if hasattr(self, '_ordered_fields') else fields
+        for field in get_fields(self):
+            value = getattr(self, field.name)
+            if value in conf.EMPTY_VALUES:
+                continue
+            if hasattr(value, 'to_dict'):
+                value = value.to_dict()
+            data.append((self._to_camel_case(field.name), self._format_field(field.name, value)))
+        return OrderedDict(data)
+
+
+@dataclass
+class CartItem(ConvertMixin):
+    """Cart item for creating card payment."""
+    # Documentation: https://github.com/csob/paymentgateway/wiki/Basic-methods#item---cart-item-object-cart-
+
+    name: str
+    quantity: int
+    amount: int
+    description: Optional[str] = None
+
+    _max_length = {"name": 20, "description": 40}
+
+
+@dataclass
+class CustomerAccount(ConvertMixin):
+    """Customer account data."""
+    # Documentation: https://github.com/csob/paymentgateway/wiki/Purchase-metadata#customeraccount-data-
+
+    created_at: Union[None, date, datetime] = None
```

Comment:
It should be defined more like this:

```python
DateType = Union[datetime, date]

class ...
    created_at: Optional[DateType] = None
```

Reply:
According to [documentation](https://docs.python.org/3/library/typing.html#typing.Optional) `Union[X, None]` is equivalent to `Optional[X]` and union of unions are flattened. `mypy` doesn't seem to complain about it either.

Reply:
With 3.10+ we will have even more expressive syntax `None | date | datetime`.

Reply:
I see this annotation as a type definition (date or datetime) and then whether it is mandatory or not. That's why I think it should be split. It's not a syntax issue.

Furthermore, I think we should only use the `datetime` type. The documentation somewhere says only date, but in the examples they have `datetime` everywhere. So I think the type should be 

`created_at: Optional[datetime] = None`.

Reply:
This definition has the advantage of simplicity.

Reply:
OK, I get your point. `date` was removed and unions with `None` avoided.